### ISSUE TITLE
guides: steps in a guide should be numbered

### ISSUE THREAD

**Issue by myitcv** [play-with-go/play-with-go#314]
@stevetraut observed that guides would be much more readable with numbered steps:

* the delivery could be much more concise (fewer filler words, more direct language)
* easier for users to reference points like "I got confused by step N"

If possible we should leverage markdown numbering here, but I fear this is a lost cause (would be awkward to maintain, manually starting a numbered list at the number you left off from before). So this will likely require some sort of `preguide` help.

Questions:

* do we have a simple step number monotonically increasing across sections, e.g. `1`, `2`..., or steps per section, e.g. `1.1`, `1.2`.... The former seems more straightforward
* would be good to be able to support some form of automatic referencing here, e.g. "this is similar to what you did back in step {{{ ref_to_prev_step }}}"

Implementation notes:

* Implementing numbering as a function, `numberStep`, in the `text/template` parser seems like it would work
* Referencing might also be possible via something like `{{{ $myref := currentStep }}}` and then later using that via `{{{ .myref }}}`, however we will need to be careful with clashes in the `Defs` of a guide

**Comment by rkbright** [comment]
I agree, some sort of standard page template may be helpful. I really like the percent complete bar, gives you a good reference for how far you've gone. Why not just start with a simple section ordering template as suggested? That would be just enough to satisfy the use case. 

I can try implementing on my fork and add you for review?

**Comment by myitcv** [comment]
Thanks @rkbright 

> Why not just start with a simple section ordering template as suggested?

I'm assuming this have people numbering steps manually, at least initially? 

Certainly having the right CSS/template would be great to kick things off, not least because that's the hardest bit to get "right".

**Comment by myitcv** [comment]
Oops, my message sent too early. 


> I can try implementing on my fork and add you for review?

This would be great, thank you.

**Comment by rkbright** [comment]
Right, I did a little googling and came across this: https://gist.github.com/patik/89ee6092c72a9e39950445c01598517a.

I'm thinking this will take care of numbering sections for all posts, and automatically at that - to your point, we do not want to do this manually.

**Comment by myitcv** [comment]
> Right, I did a little googling and came across this: https://gist.github.com/patik/89ee6092c72a9e39950445c01598517a.

That looks great!

> I'm thinking this will take care of numbering sections for all posts, and automatically at that - to your point, we do not want to do this manually.

Indeed. 

What about references? Is there any way to make them work with this approach? If not, I guess it's not out of the question to handle references separately (effectively implementing the same logic for numbering on top of a markdown parser).

**Comment by myitcv** [comment]
@rkbright - quick follow up question. Generally speaking it won't be headings that themselves will be the next step. Rather the paragraphs within them that need to be numbered. What's the best way of handling this?

**Comment by marcosnils** [comment]
:wave: in play-with-docker we added a "Course index" menu to help with this problem. It's very useful as well because it allows having direct links to the different sections of the guide. ref: https://birthday.play-with-docker.com/pods-with-docker/#launch-docker-in-docker

I recall we decided to remove this functionality in PWG Paul, I can't remember what was the specific reason. Maybe we can think about bringing that back?

**Comment by rkbright** [comment]
@myitcv - I was thinking we can explore the use of named anchors to reference the sections throughout the document. As part of the use case, are you also thinking references would span one or more posts? 

> Example: https://stackoverflow.com/questions/6695439/how-to-link-to-a-named-anchor-in-multimarkdown

re: you're other question.

> Generally speaking it won't be headings that themselves will be the next step. Rather the paragraphs within them that need to be numbered. What's the best way of handling this?

What if we stick with the same convention in the gist [above](https://gist.github.com/patik/89ee6092c72a9e39950445c01598517a). The sections would use the convention 1.1, 1.1.1, 2.1, 2.1.1. I agree it's not the cleanest implementation and somebody may have something better, but it does satisfy our use case.

I see @marcosnils just posted a comment, the index in the play-with-docker link looks good 👍!

**Comment by myitcv** [comment]
There's a few different things I think we're trying to solve for when it comes to guide structure (please let me know if I miss something):

1. giving the user a good overview of the guide
2. allowing the user to be clear where they are in the guide - heading/subheadings help in this respect, particular if the guide is long
3. allowing the user refer to a specific step in case of problems understanding etc

Point 1 is meant to be covered by the preamble/introduction at the top of each guide. A course index might help a bit in this respect, but it doesn't really make sense for a user to jump to a later section until we can restore sessions.

Point 2 is, hopefully, already covered by the headings we have, combination with the introduction.

Point 3 is, I think, what we're trying to solve for in this issue. 

The example I have in mind when referring to number paragraphs is the work @stevetraut has done for the official Go docs on modules (https://golang.org/doc/tutorial/create-module is one such example). This allows the user to be precise in what they are referring to. Numbering headings I think we:

* end up with too many headings (there will be many more steps than headings)
* don't give the user the granularity of reference

Hence why, @rkbright, I'm not sure using headings is going to work. 

We're also trying to solve for being able to refer to previous/named points as well, and have the reference numbers (with links?) be guaranteed to be correct.

All that said, I'm not clear how to actually implement this. But hopefully that helps to explain where my thinking is coming from. 

Would welcome your thoughts! Obviously I'm very open to being swayed to other approaches.

**Comment by rkbright** [comment]
Thanks for the additional context @myitcv. Yeah, headers may not be the way to go here given its limitations. 

> (https://golang.org/doc/tutorial/create-module is one such example).

Looks they they used html to implement the referencing you're looking for, and leveraged lists to provide subsection outlines. I think you want to keep the posts in markdown, right? I'll keep research other options. 

> Example: https://github.com/golang/website/blob/master/_content/doc/tutorial/create-module.html

**Comment by myitcv** [comment]
> I think you want to keep the posts in markdown, right?

Yes indeed.

One thing to bear in mind is that the markdown guides themselves run through a pre-processor called [`preguide`](https://github.com/play-with-go/preguide). So we're not limited by markdown per se. Indeed you can see that we're using `text/template` references within guide markdown:

https://github.com/play-with-go/play-with-go/blame/1aadfc00102bfd3cfc81fec2ca9647e43be66a74/guides/2020-11-12-working-with-private-modules/en.markdown#L32

Here `{{{ step "public_initial_commit" }}}` is a custom `text/template` function. 

The reason I give that background is that we effectively have the full programming capability of Go at our disposal here as a result. We're not constraint to markdown/HTML. If you want to propose a scheme that uses custom function like `{{{ step }}}` that will totally work.

> I'll keep research other options.

Splendid. Thank you.
